### PR TITLE
Always show details link

### PR DESCRIPTION
**Pull request by Mikk3lRo**
Adds the view-details link permanently as requested in https://github.com/YahnisElsts/plugin-update-checker/issues/145

### CODE REVIEW

File: Puc/v4p2/Plugin/UpdateChecker.php [hunk 1]
```diff
@@ -86,6 +86,7 @@ protected function installHooks(){
 			//Override requests for plugin information
 			add_filter('plugins_api', array($this, 'injectInfo'), 20, 3);
 
+			add_filter('plugin_row_meta', array($this, 'addViewDetailsLink'), 10, 2);
```
Comment (YahnisElsts) [inline Puc/v4p2/Plugin/UpdateChecker.php]: - If you change 2 to 3 here, the callback will receive an additional argument that contains the plugin header data. That's useful for getting the plugin name.
- Remember to remove this hook in the `removeHooks` method.

File: Puc/v4p2/Plugin/UpdateChecker.php [hunk 2]
```diff
@@ -399,6 +400,49 @@ public function addCheckForUpdatesLink($pluginMeta, $pluginFile) {
 		}
 
 		/**
+		 * Add a "View Details" link to the plugin row in the "Plugins" page. By default,
+		 * the new link will appear after the "Visit plugin site" link.
```
Comment (YahnisElsts) [inline Puc/v4p2/Plugin/UpdateChecker.php]: Actually, I think this link would look better if it was *before* the "Visit plugin site" link. It's about consistency. For a plugin from wordpress.org, meta entries are in this order: 
- version
- author link
- view details
- everything else

Ideally, we should keep the same general order:
- version
- author link
- view details
- visit plugin site
- everything else

Maybe you could search $pluginMeta for a link to $pluginHeader['PluginURI'] (if set) and insert "View details" before that?

File: Puc/v4p2/Plugin/UpdateChecker.php [hunk 3]
```diff
@@ -399,6 +400,49 @@ public function addCheckForUpdatesLink($pluginMeta, $pluginFile) {
...
+		 */
+		public function addViewDetailsLink($pluginMeta, $pluginFile) {
+			$isRelevant = ($pluginFile == $this->pluginFile)
+				|| (!empty($this->muPluginFile) && $pluginFile == $this->muPluginFile);
+
+			if ( $isRelevant && $this->userCanInstallUpdates() ) {
+				//Check if view-details is already among the links (because there is an update)
+				$viewDetailsExists = false;
+				foreach ($pluginMeta as $existingMeta) {
+					if (strpos($existingMeta, 'tab=plugin-information') !== false) {
```
Comment (YahnisElsts) [inline Puc/v4p2/Plugin/UpdateChecker.php]: I'm not convinced that checking for "tab=plugin-information" is reliable enough. In theory, a plugin could have a link to a product page that or something similar that happens to include this query parameter.

How about this: Check if `$pluginData['slug']` is set. If WordPress knows the slug, it will automatically display the "View details" link. Of course, that could change in future WP versions, but this feature will need maintenance anyway to stay consistent with the core UI.

Also, please add spaces inside the outer parentheses, i.e.:
```php
( strpos($existingMeta, 'tab=plugin-information') !== false )
```

File: Puc/v4p2/Plugin/UpdateChecker.php [hunk 4]
```diff
@@ -399,6 +400,49 @@ public function addCheckForUpdatesLink($pluginMeta, $pluginFile) {
...
+		 * the new link will appear after the "Visit plugin site" link.
+		 *
+		 * You can change the link text by using the "puc_view_details_link-$slug" filter.
+		 * Returning an empty string from the filter will disable the link.
+		 *
+		 * @param array $pluginMeta Array of meta links.
+		 * @param string $pluginFile
+		 * @return array
+		 */
+		public function addViewDetailsLink($pluginMeta, $pluginFile) {
```
Comment (YahnisElsts) [inline Puc/v4p2/Plugin/UpdateChecker.php]: As mentioned above, I recommend adding another argument so that you can get the plugin header:
```php
addViewDetailsLink($pluginMeta, $pluginFile, $pluginData = array())
```

File: Puc/v4p2/Plugin/UpdateChecker.php [hunk 5]
```diff
@@ -399,6 +400,49 @@ public function addCheckForUpdatesLink($pluginMeta, $pluginFile) {
...
+					if (strpos($existingMeta, 'tab=plugin-information') !== false) {
+						$viewDetailsExists = true;
+					}
+				}
+				$linkText = apply_filters(
+					$this->getUniqueName('view_details_link'),
+					__( 'View details' )
+				);
+				if ( !empty($linkText) && !$viewDetailsExists ) {
+					$pluginTitle = $this->getPluginTitle();
```
Comment (YahnisElsts) [inline Puc/v4p2/Plugin/UpdateChecker.php]: `getPluginTitle()` could be slow since it has to hit the disk. Consider using `$pluginData['Name']` instead.

File: Puc/v4p2/Plugin/UpdateChecker.php [hunk 6]
```diff
@@ -399,6 +400,49 @@ public function addCheckForUpdatesLink($pluginMeta, $pluginFile) {
...
+					//Show the link using the same method WP does
+					$pluginMeta[] = sprintf( '<a href="%s" class="thickbox open-plugin-details-modal" aria-label="%s" data-title="%s">%s</a>',
+						esc_url( network_admin_url( 'plugin-install.php?tab=plugin-information&plugin=' . $this->slug .
+							'&TB_iframe=true&width=600&height=550' ) ),
```
Comment (YahnisElsts) [inline Puc/v4p2/Plugin/UpdateChecker.php]: `$this->slug` is not guaranteed to be URL/HTML-safe. There's no good reason to put special characters in the slug, but it's not actually impossible. So consider using `add_query_arg()` to add the `plugin` argument.

**Comment by YahnisElsts** [comment]
All right, I finally got a moment to review this properly. See my comments above. It's mostly minor stuff.

**Comment by Mikk3lRo** [comment]
I agree with all your minor change suggestions, and they should be implemented now... I did run into one that turned out to be somewhat major though:

I don't use the "View Plugin Site" link in any of my current plugins, so I hadn't thought about the order much (just copied and adapted the comment from `addCheckForUpdatesLink`).

But digging around in core again it seems that there is actually no way for a native plugin to show both - so I don't think PUC should by default either. If there are plugin details AND the user can install plugins the plugin details are shown INSTEAD of the "Visit plugin site" link.

I've modified the function to mirror this behaviour by default.

...but I imagine a lot of people will want both (including @greenhodge in https://github.com/YahnisElsts/plugin-update-checker/issues/145). I personally would prefer both if I ever make a publicly available plugin again - but for me the plugin site should come before the details (the order in core is undefined since one always excludes the other)

So anyway, long story short, I've added another filter that allows the dev to manipulate the position relative to the "visit plugin site" link using `first|last|append|replace`. `replace` being the default.

I have to say it feels somewhat more hacky now, as the only way I could imagine to detect the site link was to `sprintf` it together and `array_search` the existing meta... Well at least it's not a huge complicated link like the other one, but who knows what might happen... I don't think it's safe to simply assume that the last link in the array is the plugin site link, though that would make it a lot less fragile.

**Comment by YahnisElsts** [comment]
> I personally would prefer both if I ever make a publicly available plugin again - but for me the plugin site should come before the details (the order in core is undefined since one always excludes the other)

As I mentioned in the related issue, I think that showing both should be the default. Losing the "visit plugin site" link is probably a drawback for commercial plugins. Let's not introduce drawbacks unless we have to.

As for the link order, I'd say that showing the "view details" link first is slightly better for usability since it's consistent with the order of links for wordpress.org plugins. Let's assume a user whose preferred language uses a left-to-right reading order, and who wants to find the "view details" link. Given a wordpress.org plugin, they'll scan the links in this order:

1. version
2. author link
3. **view details**

People probably install more plugins from wordpress.org than from other places. So they would be used to this order, and would expect the "view details" link to immediately follow the author link and everything else to come afterwards.

In short, I still think "first" should be the default. And it would probably make sense to call it "before" instead of "first" since it doesn't actually make the link first in the list.

> I have to say it feels somewhat more hacky now, as the only way I could imagine to detect the site link was to sprintf it together and array_search the existing meta...

How about searching for items that contain the Plugin URI? That would be a little more robust since it would continue work even if core developers decided to change the HTML code of the link (like adding new classes or attributes).

**Comment by Mikk3lRo** [comment]
You've convinced me that "View details" *should* come first if both links are shown. I still kinda disagree to make the default different from core (by showing both links by default)... but you're the boss here, and I don't feel particularly stronly against it ;)

I've renamed `first` and `last` to `before` and `after` and made `before` the default instead of `replace`.

The "Visit plugin site" detection is changed to look for anything containing a link to the `PluginURI` even if this link gets classes, a title or anything like that added in the future:

    preg_match('#<a[^>]*\shref="' . $pluginData['PluginURI'] . '"#', $existingLink)

My IDE has it's own ideas about how the code should be formatted... I've tried to manually force the standard you're using - I hope I caught everything... Are you using anything specific to check for consistency? (maybe I could add it in my end to make future contributions a bit smoother)

**Comment by YahnisElsts** [comment]
All right, I'll merge this now.

> My IDE has it's own ideas about how the code should be formatted... I've tried to manually force the standard you're using - I hope I caught everything... Are you using anything specific to check for consistency? (maybe I could add it in my end to make future contributions a bit smoother)

There are still some minor spacing issues, but I'll fix those myself. I use code style settings in PhpStorm. It has a keyboard shortcut for reformatting code. If you like, I can export my settings to XML and post them somewhere, but you'll  need PhpStorm or another JetBrains IDE to actually use them.
